vignettes/query.Rmd

```diff
@@ -0,0 +1,212 @@
+
+For example, the query
+
+```r
+outpack_query(quote(parameter:x == 1), scope = quote(name == "data"))
+```
+
+is equivalent to
+
+```r
+outpack_query(quote(parameter:x == 1 && name == "data"))
+```
+
+`orderly2` uses this functionality when resolving dependencies with `orderly2::orderly_dependency`.
+
+### Scoping on name
+
+Very often users will want to scope by name so instead of passing `scope` argument there is a shorthand `name` argument for usability.
+
+```r
+outpack_query(quote(parameter:x == 1), name = "data")
+```
+
+Which is the equivalent of
+
+```r
+outpack_query(quote(parameter:x == 1), scope = quote(name == "data"))
+```
+
+## Dependencies
+
+If we have 2 packets, where B depends on output from A (i.e. we call `outpack_packet_use_dependency(id_a, ...)` when running packet B) we can draw this as.
+
+```{r, echo = FALSE, fig.align = 'center'}
+knitr::include_graphics("images/dependency_terminology.png")
+```
+
+We could equivalently say
+
+* A is a dependency of B, B depends on A
+* A is used by B , B uses A
+* A is a parent of B, B is a child of A
+
+With the tree of dependencies built by outpack we might want to search for packets which have been used by another packet. We can use the query function `usedby(id)` to list all packets which are used by `id`. This will search recursively through all packets used by `id` and its parents and its parents parents and so on.
+
+The optional second arg `immedate` is `FALSE` by default, if set to `TRUE` then we search only for immediate (e.g. level 1) dependencies.
+
+Being able to search through dependencies like this means if we have some packet structure like
+
+```{r, echo = FALSE, fig.align = 'center'}
+knitr::include_graphics("images/dependency_recursive.png")
+```
+
+and we want to know the `id` of `A` which was used by `C` we can find this using `outpack_search`
+
+```r
+outpack_search(quote(usedby(latest(name == "C"))), name = "A")
+```
+
+`usedby` can be combined with groupings and scope:
+
+```r
+outpack_search(quote(usedby(latest(name == "C")) && parameter:year == 2022),
+               name = "A")
+```
+
+The depth that `usedby` will recurse can be controlled by setting the `depth` e.g.
+
+```r
+outpack_search(quote(usedby(latest(name == "C"), depth = 1)), name = "A")
+```
+
+will search for just immediate parents of `C`. `depth` can be any positive integer, by default `depth` will recurse until it finds all parents.
+
+`usedby` can be simplified by using subqueries. Subqueries are denoted by curly braces `{}` and can either be named and passed in `subquery` arg or can be anonymous. The query below is equivalent to the above but uses a subquery for `C`.
+
+```r
+outpack_search(quote(usedby({C}) && parameter:year == 2022),
+               name = "A",
+               subquery = list(C = quote(latest(name == "C"))))
+```
+
+There are two important things to note about `usedby`:
```

Comment:
I *think* you might need a newline before the `*` - this hasn't rendered as a bullet for me with devtools::build_vignettes()